connectionpool.go

```diff
@@ -719,76 +754,26 @@ func (pool *hostConnPool) fill() {
 	// determine the filling work to be done
 	startCount := len(pool.conns)
 	fillCount := pool.size - startCount
+	pool.mu.RUnlock()
 
 	// avoid filling a full (or overfull) pool
 	if fillCount <= 0 {
-		pool.mu.RUnlock()
 		return
 	}
 
-	// switch from read to write lock
-	pool.mu.RUnlock()
-	pool.mu.Lock()
-
-	// double check everything since the lock was released
-	startCount = len(pool.conns)
-	fillCount = pool.size - startCount
-	if pool.closed || pool.filling || fillCount <= 0 {
-		// looks like another goroutine already beat this
-		// goroutine to the filling
-		pool.mu.Unlock()
-		return
-	}
-
-	// ok fill the pool
-	pool.filling = true
-
-	// allow others to access the pool while filling
-	pool.mu.Unlock()
 	// only this goroutine should make calls to fill/empty the pool at this
 	// point until after this routine or its subordinates calls
 	// fillingStopped
 
-	// fill only the first connection synchronously
-	if startCount == 0 {
+	for fillCount > 0 {
 		err := pool.connect()
+		// TODO(zariel): if err != nil mark host DOWN so that we dont try to connect again
```

Comment:
No, its so that once we have the concept of hosts being UP or DOWN (from server events) we can detect that a host is not reachable and mark it such that it wont be used for queries.